src/pybind/mgr/cephadm/module.py

```diff
@@ -700,12 +700,16 @@ def _do_upgrade(self):
                     r = json.loads(''.join(out))
                     if r.get('image_id') != target_id:
                         self.log.info('Upgrade: image %s pull on %s got new image %s (not %s), restarting' % (target_name, d.nodename, r['image_id'], target_id))
-                        self.upgrade_state['image_id'] = r['image_id']
+                        self.upgrade_state['target_id'] = r['image_id']
                         self._save_upgrade_state()
                         return None
 
                 self._update_upgrade_progress(done / len(daemons))
 
+                if d.status <= 0:
+                    if d.container_image_name == target_name:
```

Comment:
Hm. do we have a way to query the image of a stopped daemon? if not, we're going to need one eventually

Reply:
yes and know.  the unit.run specifies the image by name,and that name is recorded in unit.image, so we report that (container_image_name).  the hash we do not report because it is effectively a runtime thing.. we won't know until the container starts what the hash for the image name will be at start time.

Reply:
Hm. we now have to deal with what's configure in the store, and what is written down into the unit.run file.

Reply:
That's basically what upgrade does... it reconciles what version/image the cluster wants with what the on-disk unit.run and running container are using.